### ISSUE TITLE
Spike: how can we tell which assignment an annotation belongs to?

### ISSUE THREAD

**Issue by seanh** [hypothesis/lms#4911]
Currently annotations are only associated with documents, not with assignments.

* If you create two assignments for the same document all the annotations created in one assignment will appear in the other assignment, because the annotations are associated with the document not with the assignments. (This is only true if the assignments are both in the same course and use the same course group, section or Canvas group. Assignments for the same document in different courses don't see each other's annotations because different courses have different groups.)
* If you edit an assignment and change its document then all the assignment's annotations disappear. Of course the annotations would no longer anchor if the assignment has been changed to a different document with different text. But you might hope that they'd still appear as orphans. In fact the annotations are completely gone because they're associated with the document not the assignment.

This is a problem for the email notifications feature if we're talking about counting numbers of new annotations in an assignment or in a course, because we don't actually know which assignment or course an annotation belongs to.

I think the solution to this probably involves adding new annotation ID URIs to annotations created in LMS assignments.

**Comment by seanh** [comment]
Actually, I don't think we're gonna need to do this.

**Consider the case of a course that has two assignments for the same document.** A student who has never annotated before launches assignment A and creates an annotation. Since the two assignments have the same document the annotation also appears in assignment B.

How many annotations does this student now have in assignments A and B? The way that Hypothesis is currently implemented the answer is that the student has 1 annotation in assignment A and one annotation in assignment B.

Which assignments has this student annotated? The way that Hypothesis is currently implemented the answer is that the student has annotated in both assignment A and assignment B.

I think the digest emails can just use the above answers. The instructor will receive an email saying that assignment A has one annotation from one student and assignment B also has one annotation from one student.

**Now consider the case of an assignment whose document changes.** A student creates an annotation in the assignment then a teacher changes the assignment's document.

How many annotations does this student now have in the assignment? The way that Hypothesis is currently implemented the answer is that the student has no annotations in the assignment.

Which assignments has this student annotated? The way that Hypothesis is currently implemented the answer is that the student has not annotated any assignments.

Again I think the digest emails can just use the above answers. The instructor's digest email will not count any new annotations from this student on the assignment.

* * *

We should also recognise that both of the above are presumably fairly rare edge cases. If they were common I assume we would have been asked to fix them by now, since the LMS app is quite broken in these cases.

* * *

We could fix "assignment tracking" as a separate issue. Adding assignment URIs to annotations (as previously discussed) and fixing it so that multiple assignments for the same document don't show each other's annotations in the client, and an assignment whose document changes doesn't lose its annotations in the client. The fix would then also apply to the data used by digest emails. But it's a separate problem.

**Comment by seanh** [comment]
Re-opening this because @marcospri and @jon-betts pointed out another issue to me:

For certain types of assignments like Canvas Files, Blackboard Files, and Brightspace Files assignments the LMS app doesn't actually have _any_ document URL for the assignment in its DB. Instead it has a custom `canvas-file://...` or `blackboard-file://...` etc URL. These custom URLs don't end up in the h DB at all and can't be used to search for annotations in h. The URLs that end up in h are the resolved download URLs of the files (which are usually temporary and possibly per-user URLs).

For this reason I think we may have to introduce unique and stable "assignment URIs" for LMS assignments after all, before we can say anything about annotations in assignments in these email digests.

**Comment by marcospri** [comment]
This is done now, we include the assignment information in the H DB, as part of annotatio_metadata.
